overlord/servicestate/quota_handlers.go

```diff
@@ -123,54 +124,118 @@ func (m *ServiceManager) doQuotaControl(t *state.Task, _ *tomb.Tomb) error {
+		// will happen regardless if we get rebooted after
+		// unlocking the state - if we got rebooted before unlocking
+		// the state, none of the changes we made to state would
+		// be persisted and we would run through everything above
+		// here again, but the second time around EnsureSnapServices
+		// would end up doing nothing since it is idempotent.
+		// So in the rare case that snapd gets restarted but is not a
 		// reboot also record which services do need
 		// restarting. There is a small chance that services
 		// will be restarted again but is preferable to the
 		// quota not applying to them.
-		if err := rememberQuotaStateUpdated(t, appsToRestartBySnap); err != nil {
+		if err := rememberQuotaStateUpdated(t, servicesAffected, refreshProfiles); err != nil {
 			return err
 		}
-
 	}
 
-	if err := restartSnapServices(st, t, appsToRestartBySnap, perfTimings); err != nil {
-		return err
+	if len(servicesAffected) > 0 {
+		var prevTaskSet *state.TaskSet
+		chg := t.Change()
+		queueTasks := func(ts *state.TaskSet) {
+			if prevTaskSet != nil {
+				prevTaskSet.WaitAll(ts)
+			} else {
+				ts.WaitFor(t)
+			}
+			chg.AddAll(ts)
+			prevTaskSet = ts
+		}
+
+		if refreshProfiles {
+			queueTasks(addRefreshProfileTasks(st, servicesAffected))
+		}
+		queueTasks(addRestartServicesTasks(st, qc.QuotaName, servicesAffected))
 	}
+
 	t.SetStatus(state.DoneStatus)
 	return nil
 }
 
+func addRefreshProfileTasks(st *state.State, servicesAffected map[*snap.Info][]*snap.AppInfo) *state.TaskSet {
+	ts := state.NewTaskSet()
+	for info := range servicesAffected {
+		setupProfilesTask := st.NewTask("setup-profiles", fmt.Sprintf(i18n.G("Update snap %q (%s) security profiles"), info.SnapName(), info.Revision))
+		setupProfilesTask.Set("snap-setup", &snapstate.SnapSetup{
+			SideInfo: &snap.SideInfo{
+				RealName: info.SnapName(),
+				Revision: info.Revision,
+			},
+		})
+		setupProfilesTask.WaitAll(ts)
+		ts.AddTask(setupProfilesTask)
+	}
+	return ts
+}
+
+func addRestartServicesTasks(st *state.State, grpName string, servicesAffected map[*snap.Info][]*snap.AppInfo) *state.TaskSet {
+	ts := state.NewTaskSet()
+
+	getServiceNames := func(services []*snap.AppInfo) []string {
+		var names []string
+		for _, svc := range services {
+			names = append(names, svc.Name)
+		}
+		return names
+	}
+
+	sortedInfos := make([]*snap.Info, 0, len(servicesAffected))
+	for info := range servicesAffected {
+		sortedInfos = append(sortedInfos, info)
+	}
+	sort.Slice(sortedInfos, func(i, j int) bool {
+		return sortedInfos[i].InstanceName() < sortedInfos[j].InstanceName()
+	})
+
+	for _, info := range sortedInfos {
+		restartTask := st.NewTask("service-control", fmt.Sprintf("Restarting services for snap %q", info.InstanceName()))
+		restartTask.Set("service-action", ServiceAction{
+			Action:   "restart",
+			SnapName: info.InstanceName(),
+			Services: getServiceNames(servicesAffected[info]),
+		})
+		restartTask.WaitAll(ts)
```

Comment:
Same issue here

Reply:
here in theory we could just have them run in parallel

Reply:
I have updated the solution, I am not running them in parallel to keep the current test setup. I think running them in parallel is out of scope for this PR and could be done in a future PR where the test setup could be changed.